antlir/bzl/shape.bzl

```diff
@@ -299,38 +299,94 @@ def _check_collection_type(x, t):
         return error
     return "unsupported collection type {}".format(t.collection)  # pragma: no cover
 
-def _shapes_for_field(field_or_type):
-    # recursively codegen classes for every shape that is contained in this
-    # field, or any level of nesting beneath
+def _nesting(type, path):
+    return struct(
+        type = type,
+        path = path,
+    )
+
+def _codegen_type_aliases(nestings):
+    return [
+        "typeof_{field} = {type}".format(
+            field="__".join(nesting.path),
+            type=_python_type(nesting.type),
+        )
+        for nesting in nestings
+    ]
+
+def _nested_types(type, path = None):
+    # recursively flatten a nesting of composed types, such that the result is
+    # in reverse topological ordering, that is, deepest first
+    basepath = path or []
+    nestings = []
+
+    # this is needed due to recursion on types inside collections
+    type = type.type if _is_field(type) else type
+
+    if _is_shape(type):
+        for name, field in type.fields.items():
+            path = basepath + [name]
+            nestings.extend(_nested_types(field.type, path))
+            nestings.append(_nesting(field.type, path))
+    elif _is_collection(type):
+        item_types = []
+        if types.is_list(type.item_type) or types.is_tuple(type.item_type):
+            item_types = list(type.item_type)
+        else:
+            item_types = [type.item_type]
+
+        for i, t in enumerate(item_types):
+            # when types are not associated to any field name, we use indexes
+            path = basepath + [str(i)]
+            nestings.extend(_nested_types(t, path))
+            nestings.append(_nesting(t, path))
+
+    return nestings
+
+def _codegen_type(type, generated):
     src = []
-    if _is_field(field_or_type):
-        field = field_or_type
-        if _is_shape(field.type):
-            src.extend(_codegen_shape(field.type))
-        if _is_collection(field.type):
-            item_types = []
-
-            # some collections have multiple types and some have only one
-            if types.is_list(field.type.item_type) or types.is_tuple(field.type.item_type):
-                item_types = list(field.type.item_type)
-            else:
-                item_types = [field.type.item_type]
-
-            for t in item_types:
-                src.extend(_shapes_for_field(t))
-        if _is_enum(field.type):
-            src.extend(_codegen_enum(field.type))
-    elif _is_shape(field_or_type):
-        src.extend(_codegen_shape(field_or_type))
+
+    # avoid duplicates
```

Comment:
I'd prefer if `_codegen_type` just always generated the code, and you moved this duplicate detection to the outer loop in `_codegen_nested_classes`

Reply:
I realized this could be made simpler.
`_codegen_nested_classes` only generates, well, nested classes, which means enums or shapes.
`_codegen_type` was a recursive helper. But since the hierarchy is already flattened, we can remove it.